### ISSUE TITLE
Inferencing with quantized model

### ISSUE THREAD

**Issue by minhhoangbui** [alibaba/MNN#904]
My system: PC-Ubuntu18.04

I succeeded to run .mnn model with Python API. However, after quantization, it doesn't work any more
Here is my json file
`{
  "format": "BGR",
  "mean": [128.0, 128.0, 128.0],
  "normal": [0.00392, 0.00392, 0.00392],
  "width": 368,
  "height": 368,
  "path": "/mnt/hdd3tb/Datasets/COCO2017/images/val2017",
  "used_image_num": 5000,
  "feature_quantize_method": "KL",
  "weight_quantize_method": "MAX_ABS"
}`

and here is my inference code
```
def infer(self, image):
        height, width, _ = image.shape
        scale = (self.input_size / width, self.input_size / height)
        scaled_image = cv2.resize(image, (0, 0), fx=scale[0], fy=scale[1],
                                  interpolation=cv2.INTER_CUBIC)
        scaled_image = (scaled_image - 128) / 255.0
        scaled_image = np.float32(scaled_image)
        scaled_image = np.expand_dims(scaled_image, axis=0)
        tmp_input = MNN.Tensor((1, self.input_size, self.input_size, 3),
                               MNN.Halide_Type_Float, scaled_image,
                               MNN.Tensor_DimensionType_Tensorflow)
        self.input_tensor.copyFrom(tmp_input)
        self.interpreter.runSession(self.session)
        self.heatmaps_tensor.copyToHostTensor(self.heatmaps_output)
        self.pafs_tensor.copyToHostTensor(self.pafs_output)
        heatmaps = np.squeeze(self.heatmaps_output.getData())
        pafs = np.squeeze(self.pafs_output.getData())
```

I wonder whether it has something to do with batch dimension in json file. As you can see, my model accepts 4D tensor as input

Secondly, why when I change the number of images used in json file, the output differs?

**Comment by jeonginseo** [comment]
try to quantize with mean(0,0,0) std(1,1,1)

**Comment by minhhoangbui** [comment]
> try to quantize with mean(0,0,0) std(1,1,1)

Can you explain why? It seems counter-intuitive

**Comment by jeonginseo** [comment]
I had same problem as you.
And I saw the quantization doc in yuque and thought the input image would have floating point range

**Comment by yyfcc17** [comment]
The preprocess step should be the same with your training code. in MNN, it preprocess the image as (image - mean) * normals. you can adjust these factors according to your own model.

but now, we suggest you use MNN python quant-aware-training to quantize your model. cause it's more easy and flexible to use. and also, you can use it without training, by remove the line
```
solver.step(loss)
```
just input preprocessed data to the model. here is the doc:
C++: https://www.yuque.com/mnn/en/tqr7ft
python: https://www.yuque.com/mnn/en/ricgpz https://github.com/alibaba/MNN/tree/master/pymnn/examples/MNNTrain/quantization_aware_training

**Comment by minhhoangbui** [comment]
@yyfcc17 I did follow your advice and here is the code that I prepared to quantize my current mnn model
```
class CocoDataset(MNN.data.Dataset):
    def __init__(self, labels, images_folder):
        super(CocoDataset, self).__init__()
        with open(labels, 'r') as fp:
            self.labels = json.load(fp)
        self.images_folder = images_folder

    def __len__(self):
        return len(self.labels['images'])

    def __getitem__(self, idx):
        file_name = self.labels['images'][idx]['file_name']
        img = cv2.imread(os.path.join(self.images_folder, file_name),
                         cv2.IMREAD_COLOR)
        img = cv2.resize(img, (368, 368))
        img = (img - 128) / 255.0
        dv = F.const(value_list=img.flatten().tolist(),
                     shape=[368, 368, 3], data_format=F.data_format.NHWC,
                     dtype=F.dtype.float)
        return [dv]


def quantize(args):
    dataset = CocoDataset(args['label'], args['images_dir'])
    data_loader = MNN.data.DataLoader(dataset, batch_size=32, shuffle=False)
    data_loader.reset()
    net = nn.load_module_from_file(args['model'], for_training=True)
    net.train(True)
    nn.compress.train_quant(net)

    for _ in range(data_loader.iter_number):
        example = data_loader.next()
        _ = net.forward(example)
    F.save(net.parameters, args['output'])
```
But I got an error at this line `example = data_loader.next()`
Do you have any idea about this

**Comment by yyfcc17** [comment]
i'm sorry, but for now, the dataset has to return a <data, target> pair, so you need to construct any const var as target and return it, like
```
dt = F.const(...)
return [dv], [dt]
```
after you changed the dataset, another error may occur in your code, is that the example need change to 
``` example[0][0]```
and also, the fed data may need to convert to NC4HW4 format, if you has error in forward later, by using 
```F.convert(data, F.NC4HW4)```

one last thing, the underlying machenism of mnn's dynamic graph use lazy expression, so you may need to 
```
pred = net.forward(data)
len(pred.read())
...
```
which means, if you don't need the predict output, the whole computation may not excute at all

**Comment by minhhoangbui** [comment]
@yyfcc17 Thanks for your quick response. I think that we're nearly there
At the moment, my code is like:
```
class CocoDataset(MNN.data.Dataset):
    def __init__(self, labels, images_folder):
        super(CocoDataset, self).__init__()
        with open(labels, 'r') as fp:
            self.labels = json.load(fp)
        self.images_folder = images_folder

    def __len__(self):
        return len(self.labels['images'])

    def __getitem__(self, idx):
        file_name = self.labels['images'][idx]['file_name']
        img = cv2.imread(os.path.join(self.images_folder, file_name),
                         cv2.IMREAD_COLOR)
        img = cv2.resize(img, (368, 368))
        img = (img - 128) / 255.0

        dv = F.const(value_list=img.flatten().tolist(),
                     shape=[368, 368, 3], data_format=F.data_format.NHWC,
                     dtype=F.dtype.float)
        dl = F.const([1], [], F.data_format.NHWC, F.dtype.int)
        return [dv], [dl]


def quantize(args):
    dataset = CocoDataset(args['label'], args['images_dir'])
    data_loader = MNN.data.DataLoader(dataset, batch_size=32, shuffle=False)
    data_loader.reset()
    net = nn.load_module_from_file(args['model'], for_training=True)
    nn.compress.train_quant(net, quant_bits=8)
    net.train(True)

    for i in range(10):
        example = data_loader.next()
        data = example[0][0]
        pred = net.forward(data)
        print(len(pred.read()))
    F.save(net.parameters, args['output'])
```
In this version, 
1) I add the label in the Dataset so it works fine now.
2) My MNN model is based on TensorFlow, so I guess I will keep the input as NHWC? I checked the `data` here, it has [32, 368, 368, 3]
3) However, if I put the whole 
```example = data_loader.next()
        data = example[0][0]
        pred = net.forward(data)
        print(len(pred.read()))
``` in a loop, it will have this error `Error for 25
Alloc buffer error for cpu backend
Segmentation fault`
It doesn't happen when this run 1 time only

**Comment by yyfcc17** [comment]
that is strange.

for your second question, if NHWC works fine, you don't need to change it. but i need to mention that, NC4HW4 is the internal format of MNN, some model need to convert input data to this format. so you may try this when you encounter with problem with ```forward```.

the suggestion i can give now is a question: have you tested the original float model using pymnn interface and make sure the result is right? like accuracy or some other measurements?

**Comment by minhhoangbui** [comment]
@yyfcc17 For your question, the original mnn model that I converted from TensorFlow works perfectly fine in terms of accuracy. I tested it in Ubuntu env with python API

**Comment by yyfcc17** [comment]
> @yyfcc17 For your question, the original mnn model that I converted from TensorFlow works perfectly fine in terms of accuracy. I tested it in Ubuntu env with python API

is it tested with this dynamic graph interface? can you provide the structure of your converted mnn model?

**Comment by minhhoangbui** [comment]
> > @yyfcc17 For your question, the original mnn model that I converted from TensorFlow works perfectly fine in terms of accuracy. I tested it in Ubuntu env with python API
> 
> is it tested with this dynamic graph interface? can you provide the structure of your converted mnn model?

I'm not sure what you mean by `dynamic graph interface`. But here is the structure of my mnn model that I use `netron` to export
![pose](https://user-images.githubusercontent.com/31468382/85386962-71e3b580-b56e-11ea-8a25-2baa773f0c86.png)

**Comment by yyfcc17** [comment]
i mean use  the forward method here but set ``` net.train(False)```, and remove the line
```
nn.compress.train_quant(net, quant_bits=8)
```
then test it using the dataset you constructed.

**Comment by minhhoangbui** [comment]
@yyfcc17 I do exactly what you told me: set `net.train(False)` and remove compress line. and it works fine. So I guess the problem lies at `nn.compress.train_quant`

**Comment by yyfcc17** [comment]
no, try further testing it with ```net.train(True)```

**Comment by minhhoangbui** [comment]
@yyfcc17 It still works

**Comment by yyfcc17** [comment]
OK, this is strange. so can you

1. with ```net.train(True)```, unroll the loop and see if it works?
2. with ```net.train(False)```, feed just one batch of data to do the quantization. after you save the model like the code in the demo, it is a quantized model already.

**Comment by minhhoangbui** [comment]
I got this now, I use `net.train(True)` and `nn.compress.train_quant(net, quant_bits=8)`. It works fine now. The reason is that this training progress consumes too much RAM (30gb).

After this pseudo-training, I got a quantized model exported by `F.save(net.parameters, args['output'])` but there are two issues:
1) the size remains almost the same (~16MB)
2) The structure of quantized model is broken, I cannot use the inference code anymore

**Comment by yyfcc17** [comment]
try saving like this:
```
# save model
file_name = 'quant.mnn'
net.train(False)
predict = net.forward(F.placeholder([1, 368, 368, 3], F.NHWC))
print("Save to " + file_name)
F.save([predict], file_name)
```

**Comment by yyfcc17** [comment]
> I'm not sure what you mean by `dynamic graph interface`. But [here](https://drive.google.com/file/d/1k1B2uslGhjkVPbex6wMFsmfQJQzu0ULn/view?usp=sharing) is the structure of my mnn model that I use `netron` to export

i noticed that there is no bn in your structure, have you tried this: https://www.yuque.com/mnn/en/tqr7ft#nSsYq , use --forTraining option to reserve bn in converted model, then use QAT or post training quantization.

**Comment by minhhoangbui** [comment]
@yyfcc17 Thanks to your help, I managed to run pseudo-training to produce quantized model without any error along the way.
From the `.pb` model, I converted to `.mnn` with `--forTraining` and then I use this function to sweep through the validation set for color calibration:
```
def quantize(args):
    dataset = CocoDataset(args['label'], args['images_dir'])
    data_loader = MNN.data.DataLoader(dataset, batch_size=8, shuffle=False)
    data_loader.reset()
    net = nn.load_module_from_file(args['model'], for_training=True)
    nn.compress.train_quant(net, quant_bits=8)
    net.train(True)

    for i in range(5):
        example = data_loader.next()
        data = example[0][0]
        output = net.forward([data])
        heatmap = np.array(output[0].read())
        pafs = np.array(output[1].read())
        print(heatmap.shape)
        print(pafs.shape)

    net.train(False)
    pred = net.forward([F.placeholder([1, 368, 368, 3], F.NHWC)])
    F.save(pred, args['output'])
```
But the result remains incorrect and fluctuates a lot if I change the number of iteration:
Correct result (Original MNN) : 2116.962, 136.31512
After 5 iters (quantized model): 2154.24, 285.28815 
After 10 iters (quantized model): 2153.51, 166.05766
After 40 iters (quantized model): 2259.8374, 493.55945
Do you have any ideas?

**Comment by yyfcc17** [comment]
what does these two number mean? it's the score of your model? or just a single result? for quantization, we can not make sure that every result remains the same with the original float model. but the overall score should be close.

**Comment by minhhoangbui** [comment]
Actually, they are sums of two output vectors, I put it there for simplicity. Yes, it should be close and stable. But it doesn't seem that way in my case. The more the number of iteration, the higher the results seem to be and it doesnt' saturate

**Comment by yyfcc17** [comment]
since there is no training introduced, the result may become unstable. i recomend that you caculate the score on test dataset, and choose a good one.
